### DISCUSSION TITLE
Any porting to STM32CubeIDE and ST7920 over hardware or software SPI?

### DISCUSSION THREAD

**Discussion by Leo928** [olikraus/u8g2#2459] [Q&A]
Any chance that there is similar or exact port to STM32CubeIDE and ST7920 over hardware or software SPI?  Many thanks

**Comment by olikraus** [comment]
You could check this page https://github.com/olikraus/u8g2/wiki/Porting-to-new-MCU-platform for generic and specific porting instructions.

**Reply by Leo928** [reply]
Many thanks.  I am following the example in [here](https://github.com/nikola-v/u8g2_template_stm32f103c8t6)

1. Searched and did not see (may missed), may I ask what is "**takeover_edge" and how to set** it in u8g2?  ST7920 is INPUT only amd MCU is output only.  As attached from data sheet, on clock FALLING edge, MCU drives out data on MOSI line.  On rising edge of clock, ST7920 samples the MOSI line.

2. Should I use 4wire_sw_spi or 3wire_sw_spi for ST7920 chip?  
Difference in 3_WIRES **check for dc, shift out is i<9 and if ( b & 256 )** ???

In 3wire code
		b = *data;
	if ( last_dc != 0 )
	  b |= 256;
	data++;
	arg_int--;
	for( i = 0; i < 9; i++ )
	{
	  if ( b & 256 )
....... 
in 4wire code
	b = *data;
	data++;
	arg_int--;
	for( i = 0; i < 8; i++ )
	{
	  if ( b & 128 )



``` C++
uint8_t u8x8_byte_3wire_sw_spi(u8x8_t *u8x8, uint8_t msg, uint8_t arg_int, void *arg_ptr)
{
  uint8_t i;
  uint8_t *data;
  uint8_t takeover_edge = u8x8_GetSPIClockPhase(u8x8);
  uint8_t not_takeover_edge = 1 - takeover_edge;
  uint16_t b;
  static uint8_t last_dc;
 
  switch(msg)
  {
    case U8X8_MSG_BYTE_SEND:
      data = (uint8_t *)arg_ptr;
      while( arg_int > 0 )
      {
	b = *data;
	if ( last_dc != 0 )
	  b |= 256;
	data++;
	arg_int--;
	for( i = 0; i < 9; i++ )
	{
	  if ( b & 256 )
	    u8x8_gpio_SetSPIData(u8x8, 1);
	  else
	    u8x8_gpio_SetSPIData(u8x8, 0);
	  b <<= 1;
	  
	  u8x8_gpio_SetSPIClock(u8x8, not_takeover_edge);
	  u8x8_gpio_Delay(u8x8, U8X8_MSG_DELAY_NANO, u8x8->display_info->sda_setup_time_ns);
	  u8x8_gpio_SetSPIClock(u8x8, takeover_edge);
	  u8x8_gpio_Delay(u8x8, U8X8_MSG_DELAY_NANO, u8x8->display_info->sck_pulse_width_ns);
	}    
      }
      break;` 

`uint8_t u8x8_byte_4wire_sw_spi(u8x8_t *u8x8, uint8_t msg, uint8_t arg_int, void *arg_ptr)
{
  uint8_t i, b;
  uint8_t *data;
  uint8_t takeover_edge = u8x8_GetSPIClockPhase(u8x8);
  uint8_t not_takeover_edge = 1 - takeover_edge;
 
  switch(msg)
  {
    case U8X8_MSG_BYTE_SEND:
      data = (uint8_t *)arg_ptr;
      while( arg_int > 0 )
      {
	b = *data;
	data++;
	arg_int--;
	for( i = 0; i < 8; i++ )
	{
	  if ( b & 128 )
	    u8x8_gpio_SetSPIData(u8x8, 1);
	  else
	    u8x8_gpio_SetSPIData(u8x8, 0);
	  b <<= 1;
	  
	  u8x8_gpio_SetSPIClock(u8x8, not_takeover_edge);
	  u8x8_gpio_Delay(u8x8, U8X8_MSG_DELAY_NANO, u8x8->display_info->sda_setup_time_ns);
	  u8x8_gpio_SetSPIClock(u8x8, takeover_edge);
	  u8x8_gpio_Delay(u8x8, U8X8_MSG_DELAY_NANO, u8x8->display_info->sck_pulse_width_ns);
	}    
      }
      break;
```

**Reply by olikraus** [reply]
The takeover edge should be the edge, where the display takes over the data.
The ST790 requires 4-wire spi, if I remember correctly.

**Reply by Leo928** [reply]
Many thanks, olikraus!  Some posts say ST7920 is "sensitive"???  Are people saying the chip only work in narrow range of signal timing (while other chip can take signal varying over wider timing)??

**Reply by olikraus** [reply]
I personally feel that this display is sensitive: If wires are too long or the timing not as expected, then the display behaves in wrong. I always felt that displays like ST7565 are more reliable. It least I had much lesser problems.

**Reply by Leo928** [reply]
On Arduino ATMEGA2560, which mode(s) can u8g2 works with ST7920?  
SPI_hardware, SPI_software, parallel 4 bit 8080 mode, parallel 8 bits 8080 mode, parallel 4 bit 6800 mode, parallel 8 bits 6800 mode???
Seem I use the wrong setting, it outputs 8 bits data on SPI_4wires_software mode.  ST7920 expects firstly, "special 8 bits sync" and then 8 bits containing d7_to_d4 and then, 8 bits containing d3_to_d0.  As page 26 on version 4 data sheet

**Reply by Leo928** [reply]
Many thanks for providing this wonderful library for use by community.

We have tested the following configuration and worked nicely.

Arduino IDE 2.3.2 on Window 11.

u8g2 library, version 2.34.22, downloaded by Arduino Library Manager

LCD Display 128x64, Monochrome, 2.8 inches diameter, ST7920 controller chip, White LED backlight, Display is WHITE characters on BLUE background

MCU board Arduino ATMEGA2560

Label on LCD display marked as LCD_xxx.  Pin label is pin of Arduino UNO (you may check for pin of ATMEGA 2560)

Power supply for LCD controller
LCD_VDD   3.3 volts
LCD_VSS    Ground

Backlight LED
LCD_A       5 volts
LCD_K       Ground 

Software SPI
U8G2_ST7920_128X64_F_SW_SPI u8g2(U8G2_R0, /* E_LCD clock=*/ 13, /* RW_LCD data=*/ 11, /* RS_LCD  CS=*/ 10, /* reset=*/ 8);
LCD_E is CLOCK to pin 13 of Arduino
LCD_RW is DATA_INTO_LCD (MOSI of Arduino) to pin 11 of Arduino
LCD_RS is CHIP_SELECT to pin 10 of Ardiono
LCD_RST is RESET to pin 8 of Arduino

Hardware SPI
U8G2_ST7920_128X64_F_HW_SPI u8g2(U8G2_R0, /* CS=*/ 10, /* reset=*/ 8);
LCD_RS is CHIP_SELECT to Arduino pin 10 
LCD_RST is RESET to pin 8 of Arduino
LCD_E is CLOCK to pin 52 of Arduino ATMEGA2560 (you may check for UNO pin)
LCD_RW is DATA_INTO_LCD (MOSI of Arduino) to pin 51 of Arduino ATMEGA2560 (you may check for UNO pin)

Parallel 8 bits 8080 mode
//U8G2_ST7920_128X64_F_8080 u8g2(U8G2_R0, 8, 9, 10, 11, 4, 5, 6, 7, /*enable=*/ 18 /* A4 */,   E /*cs=*/ U8X8_PIN_NONE, 
/*dc/rs=*/ 17 /* A3 */,   RS /*reset=*/ 15 /* A1 */);  	// Remember to set R/W to 0 

LCD_DB7 to Arduino pin 8
LCD_DB6 to Arduino pin 9
LCD_DB5 to Arduino pin 10
LCD_DB4 to Arduino pin 11
LCD_DB3 to Arduino pin 4
LCD_DB2 to Arduino pin 5
LCD_DB1 to Arduino pin 6
LCD_DB0 to Arduino pin 7

LCD_E is ENABLE to Arduino pin 18
LCD_RS is CHIP_SELECT to Arduino pin 17
LCD_RST is RESET to Arduino pin 15
LCD_RW is READ/nWRITE to Ground  (Ground is WRITE for nWrite line, negative_WRITE)

**Reply by olikraus** [reply]
Thanks for testing all these modes.
